cupy/linalg/einsum.py

```diff
@@ -18,22 +18,33 @@ def calc_single_view(ioperand, subscript):
             more than once, calculate diagonal for those axes.
     """
 
-    assert ioperand.ndim == len(subscript)
+    if '@' not in subscript:
+        assert ioperand.ndim == len(subscript)
+    else:
+        assert ioperand.ndim >= len(subscript.replace('@', ''))
 
-    labels = set(subscript)
+    subscripts_excluded_at = subscript.replace('@', '')
+    labels = set(subscripts_excluded_at)
     label_to_axis = collections.defaultdict(list)
     for i, label in enumerate(subscript):
         label_to_axis[label].append(i)
 
     result = ioperand
     count_dict = collections.Counter(subscript)
+    ellipsis_pos = subscript.find('@')
+
     for label in labels:
         if count_dict[label] == 1:
             continue
         axes_to_diag = []
-        for i, char in enumerate(subscript):
+        for i, char in enumerate(subscripts_excluded_at):
             if char == label:
-                axes_to_diag.append(i)
+                if ellipsis_pos == -1 or i < ellipsis_pos:
+                    axes_to_diag.append(i)
+                else:
+                    axes_to_diag.append(i - len(subscripts_excluded_at))
+        axes_to_diag = numpy.core.numeric.normalize_axis_tuple(axes_to_diag,
```

Comment:
`normalize_axis_tuple` was added v 1.13.0. We need to support older versions. Please reimplement it.